docs/design_docs/function_abi.md

```diff
@@ -1,51 +1,157 @@
+
+### Extended Type Calling Conventions
+
+While the above features of the native ABI may be sufficient for direct use by various programs, many programs and callers will need to represent various higher level types, consistently mapping them to the above facilities. This section describes calling conventions for various higher level types which do not map 1:1 to the above. Not all source language types are representable, and extending these calling conventions (and the fundamental types above) is demand driven.
+
+All of these calling conventions presume that the arity of the arguments/results of the raw function matches the user-level function, meaning that the calling convention is specified per argument/result. Higher level, whole function, transformations may also exist for some domains but are outside of the scope of this specification.
+
+#### Structure
+
+A `Structure` is a common enough entity to have a dedicated calling convention. In C-like languages, this may just be a `struct`. In Python, it is typically a `dict` with an associated schema providing a name and type bound for each of its slots. In both, its slots are of fixed arity.
+
+In this convention, such a structure is represented as a `Tuple` in the native calling convention (i.e. `!iree.list` of variant type). The order of the elements of the tuple are the natural order of the structure, where that is either:
+
+- For a C-like system where order is determinate, it is the order of declaration.
+- For a name-based system (i.e. bind to `dict`) where no order is defined, the natural order will be the lexically sorted order of the keys.
+
+#### Encoded String
+
+Most languages interop between byte arrays (i.e. the native ABI `String` type) by way of applying an encoding. Such strings are just represented as a `ByteString` in the native ABI (i.e. `!iree.list<i8>`).
+
+#### Typed List
+
+High level lists which all share the same type bound are represented as a `TypedList` in the native ABI.
+
+#### NDArray of Reference Types
+
+NDArrays of reference types are considered separately from those of value types. Internally, the code generated for them is completely different from what gets generated for numeric based arrays (i.e. has ref-counting, ownership semantics, non-POD, etc). These types are permitted for completeness, not necessarily performance: by nature they are already indirected and have overheads.
+
+In the native ABI, these are represented as a composite tuple type (i.e. today a list since sugar for tuple is not yet defined): `!iree.tuple<!iree.list<T>, !iree.list<index>>`. The first element of the tuple is the list of values, packed with a C-Layout and the second element is the list of dimension sizes.
+
+#### Reflection
+
+Additional reflection metadata may be encoded in a custom JSON form, providing additional typing hints for arguments and results. If present, this will be a reflection attribute with key `d`, containing a serialized JSON object.
```

Comment:
(if this is just for python then we can just move this entire part to the python binding docs and ignore the other bindings - above comment is more for if this is a part of the general compiler infra)

Reply:
Ok, there is a lot here and probably easier to discuss. TBH - I added all of these reflection specifics just before sending out because I realized I had forgotten and wanted to get a sketch in place to make sure I wasn't missing something. So request: can you review the rest of the code and we can iterate on the reflection bits?

Some high level thoughts, though...

What is really being modeled here (and happens to be what TF needs, if you remove the TF-isms) is the definition for calling conventions of fixed arity lists and structures (including nesting). As in a C-based language, you need some separate oracle to really do anything with those things: in C, the oracle is the header file, and in C++ it is also encoded in the mangled name. I've chosen a calling convention here such that:

a) If you just know the order and nesting of things, you can probably yolo it and carefully stuff arguments/results and get the desired results. For arbitrarily complicated nestings, this isn't going to be awesome in the same way that meticulously calling a struct-laden C-function without the header (i.e. via an FFI) is not a fun thing to be doing.
b) Should suffice for generating any language specific stub which exposes the nesting in a way that is natural for the language (i.e. a struct/class in C/C++/Java, a pytree nest of dicts/lists in python, a blessed type glob hash in perl, etc), generating code for a natural API for the entry-point. Once such stub code is generated for a module, there is no further use for the reflection metadata (it can be stripped).
c) For fully dynamic dispatch in languages that support that (really Python is the only one I care about). For this, I would require a non-stripped module in order to do the dynamic dispatch, although the bindings would work in a degraded fashion with a stripped module but would more be the Python equivalent of (a) (i.e. structs would just be lists of fields without keys, etc). I'm willing to make this concession for Python specifically because if using it dynamically in this mode, we have some more leeway to assume unstripped assets (often JIT'ed).

For (a), obviously no reflection data is needed at runtime. For (b), reflection data would be needed to generate high fidelity stubs in the target language (although simple, type-erased stubs could be generated for stripped modules if a degraded experience is ok). For (c) the reflection metadata would be needed. For that case, I haven't decided if I'll implement runtime support or just dynamically generate the Python stub as I would in (b) and eval it into the session (it's something we would need to do anyway and would reduce the per-invocation and runtime code overhead a lot).

I'll leave the precise modeling of this to a f2f (i.e. your points about early/late/etc), but I think the above is defensible and has a lot of precedent. My main goal is to not invent yet another string encoded format if I can shove some easily digestible JSON in the right places, since that will make a lot of the rest of the work a lot easier.

I think my main point is that when viewed from this perspective, it is not just "a TF" or "a Python" thing: I think I've boiled this down to some fairly universal constructs which we could expect anything to conform to if they are modeling an invocation based on ndarrays/strings/structs/primitives/fixed-sequences. The fact that I was able to twist the TF ABI into a more natural state like this is a proof that even something exotic like that does boil down to these basic concepts -- and I argue that they are worth modeling. I'd like the job of the reflection metadata to be for generating fully-typed host language stubs, of which the degenerate case of generating them on the fly for Python JIT scenarios will require an unstripped module at runtime (for full features).

As for early/late, I'm open to feedback here. I did it early for the same reason that a C/C++/Java compiler does it early: that is the point when we have the end-user definition of what the function is and assign the calling convention. I get separating the per arg/result bits early and then re-assembling, though -- just can't see the use case where that will happen yet.

Reply:
SGTM - I think that fits my mental model - having this information in the compiler is good, and having target binding lowerings be able to use it is good, and it is just the details as to how it's represented in the compiler that is the thing to work out.

Reply:
> just can't see the use case where that will happen yet.

I'll remind you of the code you wrote that breaks apart the information and then reassembles it - that I would absolutely love to delete if you think it's not needed :)

https://github.com/google/iree/blob/b7381623652bdcfd67ad2cf836f503c9d634c567/iree/compiler/Dialect/HAL/Transforms/PublicAbiGeneration.cpp

Reply:
> I'll remind you of the code you wrote that breaks apart the information and then reassembles it - that I would absolutely love to delete if you think it's not needed :)
> 
> https://github.com/google/iree/blob/b7381623652bdcfd67ad2cf836f503c9d634c567/iree/compiler/Dialect/HAL/Transforms/PublicAbiGeneration.cpp

Lol, yes I remember that and am trying not to repeat history :)

Reply:
as discussed, packed it all into the blob/json and called it `iree.abi`.